benchmarks/benchmarks/sparse_linalg_svds.py

```diff
@@ -1,31 +1,62 @@
 
+# ensure that we are benchmarking a consistent outcome;
+# (e.g. if the code wasn't able to find a solution accurately
+# enough the timing of the benchmark would become useless).
+msg = ("the benchmark code did not converge as expected, "
+       "the timing is therefore useless")
+
 
 class BenchSVDS(Benchmark):
-    # Benchmark SVD using the MatrixMarket test matrices recommended by the
+    # Benchmark SVDS using the MatrixMarket test matrices recommended by the
     # author of PROPACK at http://sun.stanford.edu/~rmunk/PROPACK/
+    # 'arpack' convergence uniformly for all singular values, while
+    # 'lobpcg' extreme singular values may converge much faster, so
+    # the assert is only evaluated on the top k/2 singular values.
+    # The assert uses the relative error since for some tested matrices
+    # the maximal singular values are very large due to poor matrix scaling.
+    # The `maxiter` and `tol` paremeters of `svds` are tuned for fair comaprison.
+    # The dense SVD solve is benchmaked as the base. 
     params = [
-        [25],
+        [20],
         ["abb313", "illc1033", "illc1850", "qh1484", "rbs480a", "tols4000",
          "well1033", "well1850", "west0479", "west2021"],
         # TODO: re-include propack
-        ['arpack', 'lobpcg']  # 'propack' failing (Aug. 2023)
+        ['arpack', 'lobpcg', 'svd']  # 'propack' failing (Aug. 2023)
     ]
     param_names = ['k', 'problem', 'solver']
 
-    def setup(self, k, problem, solver):
+    def __init__(self):
         dir_path = os.path.dirname(os.path.realpath(__file__))
         datafile = os.path.join(dir_path, "svds_benchmark_files",
                                 "svds_benchmark_files.npz")
-        matrices = np.load(datafile, allow_pickle=True)
-        self.A = matrices[problem][()]
+        self.matrices = np.load(datafile, allow_pickle=True)  
+
+    def setup(self, k, problem, solver):
+        if solver == 'svd' and problem == "tols4000":
+            # skip: fails breaking the benchmark
+            raise NotImplementedError()
+
+        self.A = self.matrices[problem][()]
+        _, s, _ = svd(self.A.toarray(), full_matrices=False)
+        self.top_singular_values = np.flip(s[:int(k/2)])
+        self.tol = k * np.max(self.A.shape) * np.finfo(float).eps
+        self.rng = np.random.default_rng(0)
 
     def time_svds(self, k, problem, solver):
-        # consider k = int(np.min(self.A.shape) * k)
-        np.random.seed(0)
-        svds(self.A, k=k, solver=solver)
+        if solver == 'svd':
+            _ = svd(self.A.toarray(), full_matrices=False)
+        else:
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                _, s, _ = svds(self.A, k=k, solver=solver, random_state=self.rng,
+                               maxiter = 50, tol=1e-6)
+            accuracy = np.sum(np.abs(1 - s[int(k/2):] / self.top_singular_values))
+            assert accuracy < self.tol, msg
```

Comment:
I think the benchmarks are a rare case where asserts are not the most visible thing.

I'm wondering if we should something like:
```suggestion
            if accuracy < self.tol:
                # failed to converge sufficently for timing to be useful
                # ensure we "destroy" performance so that benchmark suite picks up on it
                time.sleep(100)
```

Reply:
That may be too harsh, but I really have no educated opinion here. Any other reviewers supporting this change?

Reply:
We should check that an `AssertionError` actually fails the test suite. I believe it might just silently be eaten, at which point, we've achieved nothing. If that happens once in a blue moon, it's not an issue, but if there's a structural change that causes the accuracy to be missed, it's possible that it will silently keep failing.

That's assuming it's possible for the benchmark to miss the accuracy without failing the test suite; if we are confident that this cannot happen, then I'm fine with the status as-is (i.e. just don't register that particular benchmark run).

Reply:
Yes, a great approach, just must check to the opposite: `accuracy > self.tol`